File: src/main/java/com/kishan/java/leetcode/juneChallenges/InvertTree.java
package com.kishan.java.leetcode.juneChallenges;

import com.kishan.java.leetcode.juneChallenges.TreeNode;

public class InvertTree {

    public static TreeNode invertTree(TreeNode treeNode) {
        if (treeNode == null) {
            return treeNode;
        }
        TreeNode tree = swapTree(treeNode);
        tree.setLeft(invertTree(tree.getLeft()));
        tree.setRight(invertTree(tree.getRight()));
        return tree;
    }

    public static TreeNode swapTree(TreeNode tree) {
        TreeNode temp = tree.getLeft();
        TreeNode left = tree.getRight();
        TreeNode right = temp;
        tree.setRight(right);
        tree.setLeft(left);
        return tree;
    }

    public static void main(String[] args) {
        TreeNode treeNode = new TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(3)), new TreeNode(7, new TreeNode(6), new TreeNode(9)));

        System.out.println("Before Inverting");
        System.out.println("root: " + treeNode.getValue());
        System.out.println("root.left: " + treeNode.getLeft().getValue());
        System.out.println("root.left.left: " + treeNode.getLeft().getLeft().getValue());
        System.out.println("root.left.right: " + treeNode.getLeft().getRight().getValue());
        System.out.println("root.right: " + treeNode.getRight().getValue());
        System.out.println("root.right.left: " + treeNode.getRight().getLeft().getValue());
        System.out.println("root.right.right: " + treeNode.getRight().getRight().getValue());

        invertTree(treeNode);

        System.out.println("\nAfter Inverting");
        System.out.println("root: " + treeNode.getValue());
        System.out.println("root.left: " + treeNode.getLeft().getValue());
        System.out.println("root.left.left " + treeNode.getLeft().getLeft().getValue());
        System.out.println("root.left.right " + treeNode.getLeft().getRight().getValue());
        System.out.println("root.right " + treeNode.getRight().getValue());
        System.out.println("root.right.left: " + treeNode.getRight().getLeft().getValue());
        System.out.println("root.right.right: " + treeNode.getRight().getRight().getValue());


        TreeNode swap = new TreeNode(4, new TreeNode(1), new TreeNode(3));

        System.out.println("\nBefore Swapping");
        System.out.println("root: " + swap.getValue());
        System.out.println("left: " + swap.getLeft().getValue());
        System.out.println("right: " + swap.getRight().getValue());

        System.out.println("\nAfter Swapping");

        swapTree(swap);
        System.out.println("root: " + swap.getValue());
        System.out.println("left: " + swap.getLeft().getValue());
        System.out.println("right: " + swap.getRight().getValue());

    }
}

File: src/main/scala/com/leetcode/scala/monthly2020/june/invertTree.scala
package com.leetcode.scala.monthly2020.june

// Problem Statement:

/*Invert a binary tree*/


// Input:

/*
4
/   \
2     7
/ \   / \
1   3 6   9
*/


// Output:

/*
4
/   \
7     2
/ \   / \
9   6 3   1
*/


object invertTree {

  def invertTree(_tree: TreeNode): TreeNode = {
    if (_tree == null) {
      return _tree
    }
    var tree = swapTree(_tree)
    if (tree.left != null && tree.left != null) {
      tree.left = invertTree(tree.left)
      tree.right = invertTree(tree.right)
    } else if (tree.right != null) {
      tree.right = invertTree(tree.right)
    } else if (tree.left != null) {
      tree.left = invertTree(tree.left)
    }
    return tree
  }

  def swapTree(_tree: TreeNode): TreeNode = {
    var tree = _tree
    val temp = tree.left
    tree.left = tree.right
    tree.right = temp
    tree
  }

  def printElementInTree(tree: TreeNode): Unit = {
    println(s"Node Parent: ${tree.value}")
    if (tree.right != null && tree.left != null) {
      println("Right Value is")
      printElementInTree(tree.right)
      println("Left Value is")
      printElementInTree(tree.left)
    } else if (tree.right != null) {
      println("Right Value is")
      printElementInTree(tree.right)
    } else if (tree.left != null) {
      println("Left Value is")
      printElementInTree(tree.left)
    }
  }

  def main(args: Array[String]): Unit = {
    var level_1_Left = new TreeNode(2, new TreeNode(1), new TreeNode(3))
    var level_1_Right = new TreeNode(7, new TreeNode(6), new TreeNode(9))
    var root = new TreeNode(4, level_1_Left, level_1_Right)
    //    var root = new com.kishan.leetcode.juneChallenges.TreeNode(2, new com.kishan.leetcode.juneChallenges.TreeNode(3, new com.kishan.leetcode.juneChallenges.TreeNode(1)))
    var root1 = new TreeNode(2, new TreeNode(1), new TreeNode(3))
    println("Before Inverting")
    printElementInTree(root)

    println("\nAfter Inverting")
    var inverted = invertTree(root)
    printElementInTree(inverted)
  }
}
